Add Llama4 support

CatherineSue opened the pull request:
<!-- Thank you for your contribution! We appreciate it. The following guidelines will help improve your pull request and facilitate feedback. If anything is unclear, don't hesitate to submit your pull request and ask the maintainers for assistance. -->

## Motivation
- Add LM support for meta-llama/Llama4 model family.
<!-- Explain the purpose of this PR and the goals it aims to achieve. -->

## Modifications
- Add conversation template support for Llama4 models @ispobock 
- Implement Llama4 language model @ch-wan 
- Implement local attention supported by FA3 @CatherineSue 
- Support tuned MoE configs @fzyzcjy 
- Benchmark @ch-wan @ispobock 
<!-- Describe the changes made in this PR. -->

## Benchmark Results

### Llama-4-Scout-17B-16E-Instruct

- Command

```
python -m sglang.launch_server --model-path meta-llama/Llama-4-Scout-17B-16E-Instruct --port 30000 --tp 8 --mem-fraction-static 0.8 --context-length 65536
lm_eval --model local-chat-completions --model_args model=meta-llama/Llama-4-Scout-17B-16E-Instruct,base_url=http://localhost:30000/v1/chat/completions,num_concurrent=128,timeout=999999,max_gen_toks=2048 --tasks mmlu_pro --batch_size 128 --apply_chat_template --num_fewshot 0
```

- Results

```
|       Tasks       |Version|    Filter    |n-shot|  Metric   |   |Value |   |Stderr|
|-------------------|------:|--------------|-----:|-----------|---|-----:|---|-----:|
|mmlu_pro           |    2.0|custom-extract|      |exact_match|↑  |0.7516|±  |0.0038|
| - biology         |    2.1|custom-extract|     0|exact_match|↑  |0.8731|±  |0.0124|
| - business        |    2.1|custom-extract|     0|exact_match|↑  |0.7782|±  |0.0148|
| - chemistry       |    2.1|custom-extract|     0|exact_match|↑  |0.8251|±  |0.0113|
| - computer_science|    2.1|custom-extract|     0|exact_match|↑  |0.7634|±  |0.0210|
| - economics       |    2.1|custom-extract|     0|exact_match|↑  |0.8211|±  |0.0132|
| - engineering     |    2.1|custom-extract|     0|exact_match|↑  |0.6594|±  |0.0152|
| - health          |    2.1|custom-extract|     0|exact_match|↑  |0.7421|±  |0.0153|
| - history         |    2.1|custom-extract|     0|exact_match|↑  |0.6430|±  |0.0246|
| - law             |    2.1|custom-extract|     0|exact_match|↑  |0.5241|±  |0.0151|
| - math            |    2.1|custom-extract|     0|exact_match|↑  |0.8394|±  |0.0100|
| - other           |    2.1|custom-extract|     0|exact_match|↑  |0.7219|±  |0.0147|
| - philosophy      |    2.1|custom-extract|     0|exact_match|↑  |0.6593|±  |0.0212|
| - physics         |    2.1|custom-extract|     0|exact_match|↑  |0.8006|±  |0.0111|
| - psychology      |    2.1|custom-extract|     0|exact_match|↑  |0.7832|±  |0.0146|
```

### Llama-4-Maverick-17B-128E-Instruct

- Command

```
python -m sglang.launch_server --model-path meta-llama/Llama-4-Maverick-17B-128E-Instruct --port 30000 --tp 8 --mem-fraction-static 0.8 --context-length 65536
lm_eval --model local-chat-completions --model_args model=meta-llama/Llama-4-Maverick-17B-128E-Instruct,base_url=http://localhost:30000/v1/chat/completions,num_concurrent=128,timeout=999999,max_gen_toks=2048 --tasks mmlu_pro --batch_size 128 --apply_chat_template --num_fewshot 0
```

- Results
 
```
|       Tasks       |Version|    Filter    |n-shot|  Metric   |   |Value |   |Stderr|
|-------------------|------:|--------------|-----:|-----------|---|-----:|---|-----:|
|mmlu_pro           |    2.0|custom-extract|      |exact_match|↑  |0.8074|±  |0.0035|
| - biology         |    2.1|custom-extract|     0|exact_match|↑  |0.9024|±  |0.0111|
| - business        |    2.1|custom-extract|     0|exact_match|↑  |0.8466|±  |0.0128|
| - chemistry       |    2.1|custom-extract|     0|exact_match|↑  |0.8613|±  |0.0103|
| - computer_science|    2.1|custom-extract|     0|exact_match|↑  |0.8415|±  |0.0181|
| - economics       |    2.1|custom-extract|     0|exact_match|↑  |0.8614|±  |0.0119|
| - engineering     |    2.1|custom-extract|     0|exact_match|↑  |0.7193|±  |0.0144|
| - health          |    2.1|custom-extract|     0|exact_match|↑  |0.7995|±  |0.0140|
| - history         |    2.1|custom-extract|     0|exact_match|↑  |0.6903|±  |0.0237|
| - law             |    2.1|custom-extract|     0|exact_match|↑  |0.6104|±  |0.0147|
| - math            |    2.1|custom-extract|     0|exact_match|↑  |0.9023|±  |0.0081|
| - other           |    2.1|custom-extract|     0|exact_match|↑  |0.7662|±  |0.0139|
| - philosophy      |    2.1|custom-extract|     0|exact_match|↑  |0.7395|±  |0.0197|
| - physics         |    2.1|custom-extract|     0|exact_match|↑  |0.8560|±  |0.0097|
| - psychology      |    2.1|custom-extract|     0|exact_match|↑  |0.8258|±  |0.0134|
```


## Checklist


- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

CatherineSue committed:
Add Llama4 support

ch-wan committed:
complete pipeline

ch-wan committed:
fix

CatherineSue committed:
add locall_attn

ch-wan committed:
load weight

fzyzcjy committed:
Merge branch 'main-upstream' into llama4

# Conflicts:
#	python/sglang/srt/layers/attention/flashattention_backend.py

fzyzcjy committed:
rm mllama4

ch-wan committed:
load experts

ch-wan committed:
load weight

fzyzcjy committed:
Revert "rm mllama4"

This reverts commit ac4cca37d3ce0400602c1440f5deec6a63a8ab62.

ch-wan committed:
Merge commit '9fd5188965867d0335d8dde357ec81b1a6880982' into pr/CatherineSue/5092

ch-wan committed:
polish code

ispobock committed:
cleanup

ispobock committed:
format

ch-wan committed:
fix norm

ispobock committed:
add conversation template

ch-wan committed:
apply_router_weight_on_input

ispobock committed:
add chat template

ispobock committed:
format

ispobock committed:
fix load

zhyncs committed:
Merge branch 'main' into llama4

ch-wan committed:
support k > 1

ispobock committed:
lint

ch-wan committed:
fix

fzyzcjy committed:
fix mlp

CatherineSue committed:
fix local_attn support

1. Adds a `use_irope` parameter to the RadixAttention class to indicate whether a layer should use local attention based on iRoPE
2. Modifies Llama4Attention to pass `use_irope=not self.nope` to RadixAttention, leveraging the existing NoPE flag
3. Updates FlashAttentionBackend.forward_extend to check for the `use_irope` flag when determining if local attention should be used
4. Simplifies local attention activation logic by directly checking `attention_chunk_size is not None` instead of using a separate flag

qingquansong commented on python/sglang/srt/layers/attention/flashattention_backend.py:
```diff
@@ -254,7 +397,26 @@ def forward_extend(
-        page_table = metadata.page_table
+        # Check if we should use local attention
+        use_local_attn = (
+            self.use_local_attention and metadata.local_attn_metadata is not None
```
Out of curiosity, is this attention backend for all layers or partial layers only so `use_local_attention` is not applied for the full attention layers? (wondering if we need layer id to decide this together if the backend is shared by the local attention layer and full attention layers) Thank you!

qingquansong commented on python/sglang/srt/layers/attention/flashattention_backend.py:
```diff
@@ -100,6 +203,14 @@ def __init__(
+        # Local attention settings
+        self.attention_chunk_size = (
+            model_runner.attention_chunk_size
+            if hasattr(model_runner, "attention_chunk_size")
```
👀 is attention_chunk_size passed in somewhere in the model runner? couldn't find it currently. Thanks!

CatherineSue replied:
just updated it in model_runner

hebiao064 reviewed (approved):
FA3 changes looks good to me

hebiao064 commented on python/sglang/srt/layers/attention/flashattention_backend.py:
```diff
@@ -254,7 +510,28 @@ def forward_extend(
+        use_local_attn = (
+            self.attention_chunk_size is not None
+            and metadata.local_attn_metadata is not None
+            and (hasattr(layer, "use_irope") and layer.use_irope)
```
just a note here to avoid confusion:

The blog mentioned: https://blog.vllm.ai/2025/04/05/llama4.html 

> Interleaved RoPE (iRoPE): Llama 4 interleaves global attention (without RoPE) with chunked local attention (with RoPE) in a 1:3 ratio. The local attention layer attends to tokens in non-overlapping chunks, significantly reducing the quadratic complexity of attention as context length scales.

But however the model released until today doesn't follow the config, all attention are local right now. 

cc @qingquansong @ch-wan

ch-wan replied:
Thank you for pointing it out. We have fixed this issue.

ch-wan committed:
minor

ch-wan committed:
fix

ch-wan committed:
format

fzyzcjy committed:
cherry pick tuning outputs

fzyzcjy committed:
cherry pick torch compile

ispobock committed:
fix template

ispobock committed:
lint

CatherineSue committed:
clean up documentation and imports

- Remove unused imports in llama4.py
- Update copyright and documentation at the top of llama4.py
